README.md
The asynchronous distribution daemon with rsync.
==========

[![License](http://img.shields.io/badge/license-BSD-brightgreen.svg)](https://github.com/vozlt/deploy/blob/master/LICENSE)

The asynchronous distribution daemon with rsync.

Table of Contents
=================

* [Summary](#summary)
* [Architecture](#architecture)
 * [Programs](#programs)
 * [Flow](#flow)
 * [Build](#build)
 * [Security](#security)
* [Dependencies](#dependencies)
* [Compatibility](#compatibility)
* [Installation](#installation)
* [Running](#running)
 * [deploy-agent](#deploy-agent)
 * [deploy-proxy](#deploy-proxy)
 * [deploy-client](#deploy-client)
* [Configuration](#configuration)
 * [Path](#path)
 * [SSL](#ssl)
 * [Rsync options](#rsync-options)
* [Examples](#examples)
 * [STATUS](#status)
 * [RLIST](#rlist)
 * [RSYNC](#rsync)
 * [RKILL](#rkill)
 * [EXEC](#exec)
* [Author](#author)

## Summary
The deploy is a asynchronous distribution daemon with rsync.
If you have a ssl option(true), then all of data interchange is encrypted.
But if you do not use rsync ssh tunnel, then the file transfer
is not encrypted.

Deploy's features are as follows:
* Asynchronous source distribution with rsync.
* Service daemon control using user defined commands.

## Architecture
### Programs
#### deploy-agent
Deploy agent daemon for synchronization or running user defined commands

#### deploy-proxy
Deploy proxy daemon to relay from deploy-client to deploy-agent.

#### deploy-client
Deploy client to connect to deploy-agent.

### Flow
```
+---------------+      +--------------+
| deploy-client | ---> | deploy-proxy | ----+
+---------------+      +--------------+     |
	   |                                    |
	   |               +--------------+     |
	   +-------------> | deploy-agent | <---+
					   +--------------+
+---------------+             |
|    rsyncd     | <-----------+
+---------------+
```

### Build
#### [deploy-client + deploy-agent]
This is the basic design for that.
```
+---------------+      +--------------+
| deploy-client | ---> | deploy-agent | ----+
+---------------+      +--------------+     |
											|
+---------------+                           |
|    rsyncd     | <-------------------------+
+---------------+
```

#### [deploy-client + deploy-proxy + deploy-agent]
This is optional design for that.
It makes it relay the command from deploy-client to deploy-agent via
deploy-proxy that is similar to the type of api server.
```
+---------------+      +--------------+
| deploy-client | ---> | deploy-proxy | ----+
+---------------+      +--------------+     |
											|
+---------------+      +--------------+     |
|    rsyncd     | <--- | deploy-agent | <---+
+---------------+      +--------------+
```

### Security
The ssl option in config file(/etc/deploy/deploy_(agent|proxy).ini)

* SSL: `false`
 * The transferring data is a plain text.

* SSL: `true`
 * The transferring data is a encrypt string.


## Dependencies
* perl
 * perl-Crypt-OpenSSL-RSA
 * perl-Crypt-OpenSSL-AES
 * perl-Crypt-CBC
 * perl-IPC-ShareLite
* rsync

## Compatibility
* perl-5.10.x ((last tested: v5.10.1)

Earlier versions is not tested.

## Installation

1. Clone the git repository.

```
shell> git clone git://github.com/vozlt/deploy.git
```

2. Build the deploy.
```
shell> ./configure
shell> make
```

3. Install the deploy.
```
shell> make install
```

## Running

### deploy-agent
```
shell> /usr/sbin/deploy-agent --help
```
```
Usage: deploy-agent [OPTIONS]

Options:
           --path=[path]                        : set server configuration file (default: /etc/deploy/deploy_agent.ini)
           --user=[user]                        : set server running user (defaut: root)
           --group=[group]                      : set server running group (defaut: root)
           --addr=[addr]                        : set server listen address (defaut: 0.0.0.0)
           --port=[port]                        : set server listen port (default: 3440)
           --backlog=[number]                   : set server listen backlog (default: SOMAXCONN)
           --max-prefork=[number]               : limits the maximum worker process number
           --max-requests-per-child=[number]    : limits the maximum requests of child
           --max-process-timeout=[number]       : limits the maximum execution time
           --socket-timeout=[number]            : limits the maximum socket wait time
           --socket-accept-filter=[on|off]      : optimizations for a protocol's listener sockets
           --log-write=[on|off]                 : set logging
           --log-error-level=[number]           : set logging level (0-7)
           --log-error-path=[path]              : set error log path
           --log-access-path=[path]             : set access log path
           --key-cipher=[string]                : set password for authentication
           --ssl-private-key=[path]             : set ssl private key for crypt
           --debug                              : running debug mode
           --help                               : this help
```
```
shell> /usr/sbin/deploy-agent
```
```
$ ps -eo user,pid,cmd --forest | grep deploy
```
```
root     28329 deploy::agent: master process
root     28330  \_ deploy::agent: worker process
```

### deploy-proxy
```
shell> /usr/sbin/deploy-prosy --help
```
```
Usage: deploy-proxy [OPTIONS]

Options:
           --path=[path]                          : set server configuration file (default: /etc/deploy/deploy_proxy.ini)
           --user=[user]                          : set server running user (defaut: root)
           --group=[group]                        : set server running group (defaut: root)
           --addr=[addr]                          : set server listen address (defaut: 0.0.0.0)
           --port=[port]                          : set server listen port (default: 3440)
           --backlog=[number]                     : set server listen backlog (default: SOMAXCONN)
           --max-prefork=[number]                 : limits the maximum worker process number
           --max-requests-per-child=[number]      : limits the maximum requests of child
           --max-process-timeout=[number]         : limits the maximum execution time
           --socket-timeout=[number]              : limits the maximum socket wait time
           --socket-accept-filter=[on|off]        : optimizations for a protocol's listener sockets
           --log-write=[on|off]                   : set logging
           --log-error-level=[number]             : set logging level (0-7)
           --log-error-path=[path]                : set error log path
           --log-access-path=[path]               : set access log path
           --key-cipher=[string]                  : set password for authentication
           --ssl-private-key=[path]               : set ssl private key for encryption between client and proxy
           --client-worker-method=[fork|thread]   : set multi-processing method (default: fork)
           --client-ipc-path=[memory|path]        : set queue path for IPC (default: file)
           --client-port=[number]                 : set the port to connect to agent (default: 3440)
           --client-thread-timeout=[number]       : limits the client's maximum execution time
           --client-socket-timeout=[number]       : limits the client's maximum socket wait time
           --client-key-cipher=[string]           : set password for agent's authentication
           --client-ssl-public-key=[path]         : set ssl public key for encryption between proxy and agent
           --debug                                : running debug mode
           --help                                 : this help
```
```
shell> /usr/sbin/deploy-proxy
```
```
shell> ps -eo user,pid,cmd --forest | grep deploy
```
```
root     28329 deploy::proxy: master process
root     28330  \_ deploy::proxy: worker process
```

### deploy-client
```
shell> /usr/sbin/deploy-client
```
```
Usage: deploy-client [OPTIONS] [COMMANDS]

Commands:
           --order=[status|rlist|rsync|rkill|exec]    : remote command

Options:
           --hosts=[hosts]                            : remote host
           --port=[port]                              : remote port
           --args=[(groups|hosts):string]             : order hosts (proxy)
           --options=[options]                        : order options
           --key-cipher=[string]                      : set password for authentication (agent)
           --api-key=[string]                         : set api key for authentication (proxy)
           --ssl-public-key=[path]                    : set ssl public key for encryption (proxy|agent)
           --ipc-path=[memory|path]                   : set queue path for IPC (default: file)
           --worker-method=[fork|thread]              : set multi-processing method (default: fork)
           --thread-timeout=[number]                  : limits the client's maximum execution time
           --socket-timeout=[number]                  : limits the client's maximum socket wait time
           --debug                                    : running debug mode
           --verbose                                  : verbose print mode
```


## Configuration

### Path
* deploy-agent
 * /etc/deploy/deploy_agent.ini
* deploy-proxy
 * /etc/deploy/deploy_proxy.ini

### SSL
Generate (private|public).pem
```
shell> openssl genrsa 2048 > private.pem
shell> openssl rsa -in private.pem -out public.pem -outform PEM -pubout
```

### Rsync options
{force_flags} is static options to generate the results of rsync.

{default_flags} is a default options.(If you set option[] in config file, then they will be removed)

They are as follows:
* force_flags
 * --stats
 * --verbose
* default_flags
 * --archive
 * --delete
 * --force

## Examples
If agent's ssl option is true, then you need to set --ssl-public-key option.
If agent's key_cipher option is set, then you need to set --key-cipher option. (Default key_cipher: 65535) 
The below --options={rsync_group_name} first be defined in config file(/etc/deploy/deploy_agent.ini) like "[rsync::group::rsync_group_name].

### STATUS
`Description:` Print the statistics about to be transferred files.

Order to deploy-agent
```
shell> deploy-client --hosts={agent_host} --port=3440 --options={rsync_group_name} --order=status --key-cipher={key}
```
```
shell> deploy-client --hosts={agent_host} --port=3440 --options=hostname::DEST,/path/to --order=status --key-cipher={key}
```

Order to deploy-proxy
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options={rsync_group_name} --order=status --key-cipher={key}
```
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options="hostname::DEST,/path/to" --order=status --key-cipher={key}
```

Results
```Json
{  
   "targethost" : {
	  "status" : {
		 "serverStatus" : 1,
		 "returnCode" : "200",
		 "content" : {
			"transferFilesSize" : "24854724",
			"transferFilesCount" : "328",
			"alreadyRsync" : "false",
			"return" : "true"
		 },
		 "returnString" : "OK",
		 "return" : 1
	  }
   }
}

```

### RLIST
`Description:` Print the list about to be transferred files.

Order to deploy-agent
```
shell> deploy-client --hosts={agent_host} --port=3440 --options={rsync_group_name} --order=rlist --key-cipher={key}
```
```
shell> deploy-client --hosts={agent_host} --port=3440 --options=hostname::DEST,/path/to --order=rlist --key-cipher={key}
```

Order to deploy-proxy
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options={rsync_group_name} --order=rlist --key-cipher={key}
```
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options="hostname::DEST,/path/to" --order=rlist --key-cipher={key}
```

Results
```Json
{  
   "targethost" : {
	  "rlist" : {
		 "serverStatus" : 1,
		 "returnCode" : "200",
		 "content" : {
			"transferDeleteFiles" : [],
			"transferFilesSize" : "24854724",
			"transferFilesCount" : "328",
			"transferNewFiles" : [
			   "created directory /path/to",
			   "file1",
			   "path/file2",
				.
				.
				.
			],
			"return" : "true"
		 },
		 "returnString" : "OK",
		 "return" : 1
	  }
   }
}
```

### RSYNC
`Description:` Run file sync.

Order to deploy-agent
```
shell> deploy-client --hosts={agent_host} --port=3440 --options={rsync_group_name} --order=rsync --key-cipher={key}
```
```
shell> deploy-client --hosts={agent_host} --port=3440 --options=hostname::DEST,/path/to --order=rsync --key-cipher={key}
```

Order to deploy-proxy
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options={rsync_group_name} --order=rsync --key-cipher={key}
```
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options="hostname::DEST,/path/to" --order=rsync --key-cipher={key}
```

Results
```Json
{  
   "targethost" : {
	  "rsync" : {
		 "serverStatus" : 1,
		 "returnCode" : "200",
		 "content" : {
			"checksum" : "e726ab0da49b1b259d3159980d4212a0",
			"return" : "true"
		 },
		 "returnString" : "OK",
		 "return" : 1
	  }
   }
}
```

### RKILL
`Description:` Stop file sync.

Order to deploy-agent
```
shell> deploy-client --hosts={agent_host} --port=3440 --options={rsync_group_name} --order=rkill --key-cipher={key}
```
```
shell> deploy-client --hosts={agent_host} --port=3440 --options=hostname::DEST,/path/to --order=rkill --key-cipher={key}
```

Order to deploy-proxy
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options={rsync_group_name} --order=rkill --key-cipher={key}
```
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options="hostname::DEST,/path/to" --order=rkill --key-cipher={key}
```

Results
```Json
{
   "targethost" : {
	  "rkill" : {
		 "serverStatus" : 1,
		 "returnCode" : "200",
		 "content" : {
			"process" : [],
			"return" : "true"
		 },
		 "returnString" : "OK",
		 "return" : 1
	  }
   }
}
```

### EXEC
`Description:` Run user defined command.

The below --options=service:nginx,restart first be defined in config file like "service:nginx = /etc/init.d/nginx %s".
The command's arguments are separated by commas(,).

Order to deploy-agent
```
shell> deploy-client --hosts={agent_host} --port=3440 --options=service:nginx,restart --order="exec" --key-cipher={key}
```

Order to deploy-proxy
```
shell> deploy-client --api-key={key} --hosts={proxy_host} --port=3441 --args={hosts|groups}:{agent_host|agent_host_group} --options=service:nginx,restart --order=exec --key-cipher={key}
```


Results
```Json
{  
   "targethost" : {
	  "exec" : {
		 "serverStatus" : 1,
		 "returnCode" : "200",
		 "content" : {
			"content" : "U3RvcHBpbmcgbmdpbng6ICAgICAgICAgWyAgT0sgIF0NClN0YXJ0aW5nIG5naW54OiAgICAgICAgIFsgIE9LICBd",
			"command" : "L2V0Yy9pbml0LmQvbmdpbnggcmVzdGFydA==",
			"return" : "true"
		 },
		 "returnString" : "OK",
		 "return" : 1
	  }
   }
}
```

## Author
YoungJoo.Kim(김영주) [<vozlt@vozlt.com>]


ChangeLog
v0.0.1  [Tue Jan 06 2015 YoungJoo.Kim <vozlt@vozlt.com>]
    * first packaged

# vi:set ft=changelog ts=4 sw=4 et fdm=marker:


src/lib/Deploy/Proxy.pm
# @file:    Proxy.pm
# @brief:
# @author:  YoungJoo.Kim <vozlt@vozlt.com>
# @version:
# @date:

package Deploy::Proxy;

use strict;
use warnings "all";

use Deploy::Boolean qw(:Boolean);
use Deploy::Onetime;
use Deploy::Server;
use Deploy::Ipcalc;
use Deploy::Config;
use Deploy::Rsync;
use Deploy::Carp;
use Deploy::Client;

use Socket;
use JSON;
use Data::Dumper;

our @ISA = qw(Deploy::Ipcalc Deploy::Config Deploy::Carp);

sub new {
    my($class, %cnf) = @_;

    my $debug = delete $cnf{debug};
    my $path = delete $cnf{path};
    my $process_name = delete $cnf{process_name};
    my $user = delete $cnf{user};
    my $group = delete $cnf{group};
    my $addr = delete $cnf{addr};
    my $port = delete $cnf{port};
    my $backlog = delete $cnf{backlog};
    my $max_prefork = delete $cnf{max_prefork};
    my $max_requests_per_child = delete $cnf{max_requests_per_child};
    my $max_process_timeout = delete $cnf{max_process_timeout};
    my $socket_timeout = delete $cnf{socket_timeout};
    my $socket_accept_filter = delete $cnf{socket_accept_filter};

    # Client (user -> proxy -> client -> agent)
    my $client_worker_method = delete $cnf{client_worker_method};
    my $client_ipc_path = delete $cnf{client_ipc_path};
    my $client_port = delete $cnf{port};
    my $client_thread_timeout = delete $cnf{client_thread_timeout};
    my $client_socket_timeout = delete $cnf{client_socket_timeout};
    my $client_key_cipher = delete $cnf{client_key_cipher};
    my $client_ssl = delete $cnf{client_ssl};
    my $client_ssl_public_key = delete $cnf{client_ssl_public_key};

    my $log_write = delete $cnf{log_write};
    my $log_error_level = delete $cnf{log_error_level};
    my $log_error_path = delete $cnf{log_error_path};
    my $log_access_path = delete $cnf{log_access_path};
    my $key_cipher = delete $cnf{key_cipher};
    my $ssl = delete $cnf{ssl};
    my $ssl_public_key = delete $cnf{ssl_public_key};
    my $ssl_private_key = delete $cnf{ssl_private_key};
    $path = '/etc/deploy/deploy_proxy.ini' unless defined $path;

    my $self =
    bless {
        debug                   => $debug,
        path                    => $path,

        user                    => $user,
        group                   => $group,
        addr                    => $addr,
        port                    => $port,
        backlog                 => $backlog,

        max_prefork             => $max_prefork,
        max_requests_per_child  => $max_requests_per_child,
        max_process_timeout     => $max_process_timeout,

        socket_timeout          => $socket_timeout,
        socket_accept_filter    => $socket_accept_filter,
        
        log_write               => $log_write,
        log_error_level         => $log_error_level,
        log_error_path          => $log_error_path,
        log_access_path         => $log_access_path,
        key_cipher              => $key_cipher,
        ssl                     => $ssl,
        ssl_public_key          => $ssl_public_key,
        ssl_private_key         => $ssl_private_key,

        client_worker_method    => $client_worker_method,
        client_ipc_path         => $client_ipc_path,
        client_port             => $client_port,
        client_thread_timeout   => $client_thread_timeout,
        client_socket_timeout   => $client_socket_timeout,
        client_key_cipher       => $client_key_cipher,
        client_ssl              => $client_ssl,
        client_ssl_public_key   => $client_ssl_public_key,
    
        server_self             => {},
        request                 => {},
        cnf                     => {},
        
        orders                  => {},
        allow_orders            => ['status', 'rlist', 'rsync', 'rkill', 'exec'],
        ps                      => 'ps',
        proxy                   => {},
        returns                 => {},

        process_name            => $process_name,
        process_id              => $$
    }, $class;

    return bless $self;
}

sub checkAllowInet {
    my $self = shift if ref ($_[0]);
    my @hosts = @{$self->{'cnf'}{'allow::hosts'}{'host'}};
    my $remote_addr = $self->{request}{headers}{REMOTE_ADDR};

    for (@hosts) {
        my $cidr = undef;
        if (/[a-z]+/i) {
            my $iaddr = Socket::inet_aton($_) || next;
            $cidr = Socket::inet_ntoa($iaddr) || next;
        } else {
            $cidr = $_;
        }
        $self->__debug(
            {
                string => (caller(0))[3] . '() line ' . __LINE__ . ': ' . 
                "checkAllowInet() => cidr:$cidr, remote_addr:$remote_addr"
            }
        );
        $self->ipCalcWithIn($cidr, $remote_addr) && return TRUE;
    }
    $self->{returns}{content} = "Your IP[$remote_addr] doesn't have permisson to access on this server.";
    return;
}

sub checkAllowMethod {
    my $self = shift if ref ($_[0]);
    my $method = $self->{request}{headers}{METHOD};
    my @methods = qw(GET POST);

    grep(/^${method}$/, @methods) || return;
}

sub checkAllowApi {
    my $self = shift if ref ($_[0]);
    my $api_key = $self->{orders}{api_key};
    my $args = $self->{orders}{args}; # (groups|hosts):(group or hosts)

    my $proxy = $self->{proxy} = $self->getHosts();

    if ($proxy->{type} eq 'hosts') {
        
        for my $host (@{$proxy->{hosts}}) {
            if (!grep(/^${host}$/, @{$self->{cnf}{api}{list}{access_hosts}{$api_key}})) {
                $self->{returns}{content} = "You don't have permisson to access on this[$host].";
                return;
            }
        }
    } elsif ($proxy->{type} eq 'groups') {
        for my $group (@{$proxy->{groups}}) {
            if (!grep(/^${group}$/, @{$self->{cnf}{api}{list}{access_groups}{$api_key}})) {
                $self->{returns}{content} = "You don't have permisson to access on this[$group].";
                return;
            }
        }
    } else {
        $self->{returns}{content} = "Syntax error, unrecognized expression: [$args]";
        return;
    }

    return TRUE;
}

sub checkAllowUrl {
    my $self = shift if ref ($_[0]);
    my $url = $self->{request}{headers}{URL};
    my @args =  grep {/\S/} split('/', $url);
    my $target = $self->{orders}{target} = $args[0];
    my $order = $self->{orders}{order} = $args[1];
    my $options = $self->{orders}{options} = ($args[2]) ? $args[2] : $self->{request}{args}{options};
    my $args = $self->{orders}{args} = ($args[3]) ? $args[3] : $self->{request}{args}{args};
    my $api_key = $self->{orders}{api_key} = ($args[4]) ? $args[4] : $self->{request}{args}{api_key};

    $self->trim($self->{orders}{target});
    $self->trim($self->{orders}{args});
    $self->trim($self->{orders}{options});
    $self->trim($self->{orders}{api_key});

    (grep(/^$order$/, @{$self->{allow_orders}}) && $target =~ /^(api|agent)$/) || return;
}

sub __client_exec {
    my $self = shift if ref ($_[0]);
    my $api_key = $self->{orders}{api_key};
    my @hosts = @{$self->{proxy}{hosts}};

    my $client = Deploy::Client->new(
        warn => $self->{warn},
        debug => $self->{debug},
        worker_method => $self->{client_worker_method},
        ipc_path => $self->{client_ipc_path},
        port => $self->{client_port},
        thread_timeout => $self->{client_thread_timeout},
        socket_timeout => $self->{client_socket_timeout},
        ssl => $self->{client_ssl},
        ssl_public_key => $self->{client_ssl_public_key},
        key_cipher => $self->{client_key_cipher},
        orders => {
            target => $self->{orders}{target},
            order => $self->{orders}{order},
            options => $self->{orders}{options},
            args => $self->{orders}{args},
            hosts => \@hosts
        }
    );

    my $results = $client->exec();
    $self->__debug(
        {
            string => (caller(0))[3] . '() line ' . __LINE__ . ": " . Dumper($results)
        }
    );
    my $json = JSON->new->utf8(1)->pretty(1);
    my $jref = {};
    for my $res (@{$results}) {
        my $host = $res->{host};
        my $server_status = $res->{server_status};
        my $content = $res->{CONTENT};
        my $order = $self->{orders}{order};
        my $return = ($res->{HTTP_CODE} == 200) ? 1 :0;

        my $json_content = ($return) ? $json->decode($content) : FALSE;

        $jref->{$host}{$order}{return} = $return;
        $jref->{$host}{$order}{serverStatus} = $server_status;
        $jref->{$host}{$order}{content} = ($json_content) ? $json_content : $content;
        $jref->{$host}{$order}{returnCode} = $res->{HTTP_CODE};
        $jref->{$host}{$order}{returnString} = $res->{HTTP_STRING};
    }

    my $json_string = $json->encode($jref);

    return $json_string;
}

sub __proxy_exec {
    my $self = shift if ref ($_[0]);
    my $content;
    my $status = 200;
    eval {
        $content = $self->__client_exec();
    };

    if ($@) {
        $self->__debug(
            {  
                string => (caller(0))[3] . '() line ' . __LINE__ . ": $@"
            }
        );
        if (!defined($content)) {
            $status = 500;
            $content = $@
        }
    }

    return(
        {
            status => $status,
            content => $content
        }
    );
}

sub getHosts {
    my $self = shift if ref ($_[0]);
    my $args = shift || $self->{orders}{args}; # (groups|hosts):(group or hosts)
    my @hosts = ();
    my @groups = ();

    my ($left, $right) = split(/:/, lc($args), 2);
    if ($left eq 'hosts') {
        @hosts = grep {/\S/} split(/,/, $right);
    } elsif ($left eq 'groups') {
        @groups = grep {/\S/} split(/,/, $right);
        for my $group (@groups) {
            push @hosts, $self->getGroupToHosts($group);
        }
    }

    return ({type => $left, hosts => \@hosts, groups => \@groups});
}

sub getAccessHosts {
    my $self = shift if ref ($_[0]);
    my $api_key = shift || $self->{orders}{api_key};
    defined $self->{cnf}{api}{list}{access_hosts}{$api_key} || return;
    return @{$self->{cnf}{api}{list}{access_hosts}{$api_key}};
}

sub getAccessGroups {
    my $self = shift if ref ($_[0]);
    my $api_key = shift || $self->{orders}{api_key};
    defined $self->{cnf}{api}{list}{access_groups}{$api_key} || return;
    return @{$self->{cnf}{api}{list}{access_groups}{$api_key}};
}

sub getGroupToHosts {
    my $self = shift if ref ($_[0]);
    my $group = shift;
    defined $self->{cnf}{api}{list}{servers}{$group} || return;
    return @{$self->{cnf}{api}{list}{servers}{$group}};
}

sub __init_api {
    my $self = shift if ref ($_[0]);
    my @keys = keys %{$self->{cnf}{api}{list}{keys}};
    my @groups = ();
    for my $key (@keys) {
        for my $group (@{$self->{cnf}{api}{list}{keys}{$key}}) {
            push @groups, grep {/\S/} split(/,/, $group);
        }
        for my $group (@groups) {
            push @{$self->{cnf}{api}{list}{access_hosts}{$key}}, @{$self->{cnf}{api}{list}{servers}{$group}}
            if defined $self->{cnf}{api}{list}{servers}{$group};
        }
        push @{$self->{cnf}{api}{list}{access_groups}{$key}}, @groups;
        @groups = ();
    }

    return $self;
}

# first reading in serverMain(config file...)
sub workerInit {
    my $self = shift if ref ($_[0]);
    my $server_self = shift if ref ($_[0]);
    $self->{server_self} = $server_self; 
    my %cnf = $self->parseIniFile($self->{path});
    $cnf{api}{list} = {$self->parseIniFile($cnf{api}{path})};
    my $process_name = $cnf{server}{process_name};
    my $user = $cnf{server}{user};
    my $group = $cnf{server}{group};
    my $addr = $cnf{server}{addr};
    my $port = $cnf{server}{port};
    my $backlog = $cnf{server}{backlog};
    my $socket_timeout = $cnf{server}{socket_timeout};
    my $socket_accept_filter = $cnf{server}{socket_accept_filter};
    my $max_prefork = $cnf{server}{max_prefork};
    my $max_requests_per_child = $cnf{server}{max_requests_per_child};
    my $max_process_timeout = $cnf{server}{max_process_timeout};
    
    my $log_write = $cnf{server}{log_write};
    my $log_error_level = $cnf{server}{log_error_level};
    my $log_error_path = $cnf{server}{log_error_path};
    my $log_access_path = $cnf{server}{log_access_path};
    my $key_cipher = $cnf{server}{key_cipher};
    my $ssl = $cnf{server}{ssl};
    my $ssl_public_key = $cnf{server}{ssl_public_key};
    my $ssl_private_key = $cnf{server}{ssl_private_key};
    
    my $client_worker_method = $cnf{client}{worker_method};
    my $client_ipc_path = $cnf{client}{ipc_path};
    my $client_port = $cnf{client}{port};
    my $client_thread_timeout = $cnf{client}{thread_timeout};
    my $client_socket_timeout = $cnf{client}{socket_timeout};
    my $client_key_cipher = $cnf{client}{key_cipher};
    my $client_ssl = $cnf{client}{ssl};
    my $client_ssl_public_key = $cnf{client}{ssl_public_key};

    $process_name = 'deploy::proxy:' unless defined $process_name;
    $user = 0 unless defined $user;
    $group = 0 unless defined $group;
    $addr = 0 unless defined $addr;
    $port = 3441 unless defined $port;
    $backlog = SOMAXCONN unless defined $backlog;
    $socket_timeout = 5 unless defined $socket_timeout;
    $socket_accept_filter = 'false' unless defined $socket_accept_filter;
    $max_prefork = 0 unless defined $max_prefork;
    $max_requests_per_child = 0 unless defined $max_requests_per_child;
    $max_process_timeout = 0 unless defined $max_process_timeout;
    
    $log_write = 'false' unless defined $log_write;
    $log_error_level = 'info' unless defined $log_error_level;
    $log_error_path = '/var/log/deploy/error.log' unless defined $log_error_path;
    $log_access_path = '/var/log/deploy/access.log' unless defined $log_access_path;
    $ssl = 'false' unless defined $ssl;

    $client_worker_method = 'fork' unless defined $client_worker_method;
    $client_ipc_path = '/tmp/.deploy_proxy_ipc' unless defined $client_ipc_path;
    $client_port = 3440 unless defined $client_port;
    $client_thread_timeout = 10 unless defined $client_thread_timeout;
    $client_socket_timeout = 5 unless defined $client_socket_timeout;
    $client_ssl = 'false' unless defined $client_ssl;

    # Server config
    $server_self->{debug} = $self->{debug} || $server_self->{debug};
    $server_self->{process_name} = defined($self->{process_name}) ? $self->{process_name} : $process_name;
    $server_self->{user} = defined($self->{user}) ? $self->{user} : $user;
    $server_self->{group} = defined($self->{group}) ? $self->{group} : $group;
    $server_self->{addr} = defined($self->{addr}) ? $self->{addr} : $addr;
    $server_self->{port} = defined($self->{port}) ? $self->{port} : $port;
    $server_self->{backlog} = defined($self->{backlog}) ? $self->{backlog} : $backlog;
    $server_self->{socket_timeout} = defined($self->{socket_timeout}) ? $self->{socket_timeout} : $socket_timeout;
    $server_self->{socket_accept_filter} =  defined($self->{socket_accept_filter}) ? $self->{socket_accept_filter} : $socket_accept_filter;
    $server_self->{max_prefork} = defined($self->{max_prefork}) ? $self->{max_prefork} : $max_prefork;
    $server_self->{max_requests_per_child} = defined($self->{max_requests_per_child}) ? $self->{max_requests_per_child} : $max_requests_per_child;
    $server_self->{max_process_timeout} = defined($self->{max_process_timeout}) ? $self->{max_process_timeout} : $max_process_timeout;
    
    $server_self->{log_write} = defined($self->{log_write}) ? $self->{log_write} : $log_write;
    $server_self->{log_error_level} = defined($self->{log_error_level}) ? $self->{log_error_level} : $log_error_level;
    $server_self->{log_error_path} = defined($self->{log_error_path}) ? $self->{log_error_path} : $log_error_path;
    $server_self->{log_access_path} = defined($self->{log_access_path}) ? $self->{log_access_path} : $log_access_path;
    $server_self->{ssl} = defined($self->{ssl}) ? $self->{ssl} : $ssl;
    $server_self->{ssl_public_key} = defined($self->{ssl_public_key}) ? $self->{ssl_public_key} : $ssl_public_key;
    $server_self->{ssl_private_key} = defined($self->{ssl_private_key}) ? $self->{ssl_private_key} : $ssl_private_key;

    $server_self->{ssl_public_key} = $self->fileRead($server_self->{ssl_public_key}) if defined $server_self->{ssl_public_key};
    $server_self->{ssl_private_key} = $self->fileRead($server_self->{ssl_private_key}) if defined $server_self->{ssl_private_key};
    
    # Agent config
    $self->{key_cipher} = defined($self->{key_cipher}) ? $self->{key_cipher} : $key_cipher;

    # Client config
    $self->{client_worker_method} = defined($self->{client_worker_method}) ? $self->{client_worker_method} : $client_worker_method;
    $self->{client_ipc_path} = defined($self->{client_ipc_path}) ? $self->{client_ipc_path} : $client_ipc_path;
    $self->{client_port} = defined($self->{client_port}) ? $self->{client_port} : $client_port;
    $self->{client_thread_timeout} = defined($self->{client_thread_timeout}) ? $self->{client_thread_timeout} : $client_thread_timeout;
    $self->{client_socket_timeout} = defined($self->{client_socket_timeout}) ? $self->{client_socket_timeout} : $client_socket_timeout;
    $self->{client_key_cipher} = defined($self->{client_key_cipher}) ? $self->{client_key_cipher} : $client_key_cipher;
    $self->{client_ssl} = defined($self->{client_ssl}) ? $self->{client_ssl} : $client_ssl;
    $self->{client_ssl_public_key} = defined($self->{client_ssl_public_key}) ? $self->{client_ssl_public_key} : $client_ssl_public_key;

    $self->{client_ssl_public_key} = $self->fileRead($self->{client_ssl_public_key}) if defined $self->{client_ssl_public_key};

    # String to flag
    $server_self->{socket_accept_filter} = ($server_self->{socket_accept_filter} =~ /^(true|on|1)$/i) ? 1 : 0;
    $server_self->{log_write} = ($server_self->{log_write} =~ /^(true|on|1)$/i) ? 1 : 0;
    $server_self->{ssl} = ($server_self->{ssl} =~ /^(true|on|1)$/i) ? 1 : 0;
    $self->{client_ssl} = ($self->{client_ssl} =~ /^(true|on|1)$/i) ? 1 : 0;

    $self->{cnf} = \%cnf;

    $self->__init_api();

    return $server_self;
}

sub workerHandlerInit {
    my $self = shift if ref ($_[0]);
    my $server_self = $self->{server_self};
    # same reference for memory
    $self->{request} = $server_self->{request};

    if ($self->{debug}) {
        my %headers = %{$self->{request}{headers}};
        my %args = %{$self->{request}{args}};
        for (keys(%headers)) {
            print("request{$_} = $headers{$_}\n");
        }
        for (keys(%args)) {
            print("args{$_} = $args{$_}\n");
        }
    }
}

sub workerHandler {
    my $self = shift if ref ($_[0]);
    my $server_self = $self->{server_self};

    $self->workerHandlerInit($server_self);

    $self->checkAllowInet() || return({status => 403, content => $self->{returns}{content}});
    $self->checkAllowMethod() || return({status => 405});
    $self->checkAllowUrl() || return({status => 404});
    $self->checkAllowApi() || return({status => 403, content => $self->{returns}{content}});

    my $order_handler = '__proxy_exec';

    return $self->can($order_handler) ? $self->$order_handler() : (status => 400, content => "${order_handler}() not found!");
}

sub main {
    my $self = shift if ref ($_[0]);
    my $daemonize = ($self->{debug}) ? FALSE : TRUE;
    my $server = Deploy::Server->new(
            worker_init => sub { $self->workerInit(@_) },
            worker_handler => sub { $self->workerHandler(@_) }
    );

    $server->serverMain({daemonize => $daemonize});
}

AUTOLOAD {
    my $self = shift if ref ($_[0]);
    my $attr = our ($AUTOLOAD);
    $attr =~ s/.*:://;
    return(status => 400, content => "$attr() not found!");
}

sub DESTROY {
    my $self = shift if ref ($_[0]);

    # thread safe
    ($$ != $self->{process_id}) && return;

    for my $class (@ISA) {
        my $destroy = $class . "::DESTROY";
        $self->$destroy if $self->can($destroy);
    }
}

1;

# vi:set ft=perl ts=4 sw=4 et fdm=marker:
